python/ray/air/execution/tracked_actor.py

```diff
@@ -0,0 +1,36 @@
+from typing import Callable
+
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class TrackedActor:
+    """Actor tracked by an actor manager.
+
+    This object is used to reference a Ray actor on an actor manager
+
+    Existence of this object does not mean that the Ray actor has already been started.
+    Actor state can be inquired from the actor manager tracking the Ray actor.
+
+    Attributes:
+        actor_id: ID for identification of the actor within the actor manager. This
+            ID is not related to the Ray actor ID.
+
+    """
+
+    def __init__(self, actor_id: int):
+        self.actor_id = actor_id
+
+    def on_start(self, callback: Callable[["TrackedActor"], None]) -> "TrackedActor":
```

Comment:
Should these be made part of the constructor instead? I feel like mutating these with `on_X` feels a bit strange? Though let me know if this is a standard pattern that I'm just unaware of.

Reply:
It's the standard pattern in asynchronous programming, e.g. javascript/jquery.

Main reason we want to use a method and not a constructor is that the developer actually never constructs these, they get the instances from the event manager and then can specify callbacks.

Reply:
The main alternative is to pass the callbacks when scheduling the objects, e.g.

```
event_manager.schedule_task(remote_fn, on_result=callback, on_error=callback)

# vs.

event_manager.schedule_task(remote_fn).on_result(callback).on_error(callback)
```

the main reason why we should keep the current proposal is that we can change callbacks later. E.g. if we gang schedule 10 training futures and one errors, we can unset the error callbacks from the other futures so we only handle one failure, etc.

Reply:
Right, my main concern here is around the guarantees and expectations around whether or not callbacks will be invoked, which I think is particularly difficult yet important for us to define clearly for the non-happy paths (race conditions!) in asynchronous programming.

In a simple scenario, with the current API what is the expected behavior if this sequence happens?

1. `tracked_task = event_manager.schedule_task(remote_fn)`.
2. An error occurs.
3. The caller calls `tracked_task.on_error(callback)`.

Is it guaranteed that the callback will be executed? Does it depend on whether the event manager catches/processes the error before or after (3)?

> if we gang schedule 10 training futures and one errors, we can unset the error callbacks from the other futures so we only handle one failure, etc.

Similarly, for this example would the expectation be that only one failure will ever need to be processed? Or is this meant to be an optimization to reduce additional calls? I feel that regardless we would still need to have handling in the consumer (e.g. in the `Barrier`) to reject/ignore the other errors if they happen before this logic is invoked.

To make this concise:
1. Should callbacks be static or dynamic?
2. If callbacks are dynamic, what are the guarantees around which callback(s) is invoked?

Personally I would prefer to start with statically defined callbacks, and if/when we do want to support dynamic callbacks, I'd really want for us to nail down the semantics before implementation.

Reply:
In your concrete example, if you call

```
tracked_task.on_error(callback)
event_manager.wait()
```

the callback will be invoked.

If you call 

```
event_manager.wait()
tracked_task.on_error(callback)
```

it won't